cgp/genome.py

```diff
@@ -267,6 +267,172 @@ def randomize(self, rng: np.random.RandomState) -> None:
+            for current_dependencies in dependencies:
+                current_dependencies.discard(current_idx)
+
+            # update set of addable nodes
+            addable_nodes = set(
+                idx
+                for (idx, item) in enumerate(dependencies)
+                if not item and self._is_hidden_region(idx)
+            )
+            # remove current_idx and all previous used idx's
+            # from addable_nodes and increment node_idx counter
+            used_idx_list.append(current_idx)
+            addable_nodes = addable_nodes.difference(used_idx_list)
+            node_idx += 1
+
+        # change the gene values such that they fit the new location of the genes
+        for gene_idx, gene_value in enumerate(dna_for_reorder):
+            region_idx = gene_idx // length_per_region
+            if self._is_hidden_input_gene(gene_idx, region_idx) or self._is_output_input_gene(
+                gene_idx
+            ):
+                try:
+                    gene_value = self._n_inputs + used_idx_list.index(gene_value)
+                except ValueError:  # means the input is from an input node -> do nothing
+                    pass
+            dna_for_reorder[gene_idx] = gene_value
+
+        # change the invalid gene values (coming from unused arities)
+        for gene_idx, gene_value in enumerate(dna_for_reorder):
+            region_idx = gene_idx // length_per_region
+            if self._is_hidden_input_gene(gene_idx, region_idx) and gene_value > region_idx:
+                # replace value that is too large by smaller value
+                permissible_values = self.determine_permissible_values_per_gene(gene_idx)
+                gene_value = rng.choice(permissible_values)
+                dna_for_reorder[gene_idx] = gene_value
+
+        self.dna = dna_for_reorder
+
+    def _determine_dependencies(self) -> List[set]:
+
+        dependencies = []
+        for region_idx in range(self._n_regions):
+            current_node_dependencies = set()
+
+            if self._is_input_region(region_idx):
+                dependencies.append(
+                    current_node_dependencies
+                )  # input regions have no dependencies
+
+            elif self._is_hidden_region(
+                region_idx
+            ):  # can assume dependencies is not empty anymore
+                # first determine arity of current operator
+                # to only consider genes which are used in current node
+                operator_idx = region_idx * self._length_per_region
+                operator_value = self._dna[operator_idx]
+                current_arity = self._primitives[operator_value]._arity
+
+                # append each input and its respective dependencies to the dependencies
+                for idx_gene in range(
+                    1, current_arity + 1
+                ):  # shift by 1 since first gene is operator
+                    input_region_idx = self._dna[operator_idx + idx_gene]
```

Comment:
the dna actually does not store a region index, but a node index, they are just identical. not sure we should mix semantics here :thinking: